### DISCUSSION TITLE
Execution option eval: false vs "Run above"

### DISCUSSION THREAD

**Discussion by werkstattcodes** [quarto-dev/quarto-cli#2643] [Q&A]
I am struggling with the following, rather basic thing:

Setting the chunk option to eval: "false" makes quarto not evaluate the chunk when rendering the document. Only the code inside the chunk is displayed. 

When working 'interactively' (?), i.e. not rendering, on a quarto document, I frequently use the command "Run Above" or "Run Cells Above". In these cases, it seems like the chunk option "eval: false" is not respected, i.e. the code in the chunk in question is nevertheless evaluated. This is particularly annoying when the code in the chunk is very time intensive. As a work-around, I always have to comment out these chunks.

In short, my question, is there a way to get chunks not evaluated when not rendering the quarto document, but simply working on it?
Many thanks.

**Comment by cderv** [comment]
Are you using RStudio IDE ? I assuming so below

> I frequently use the command "Run Above" or "Run Cells Above". In these cases, it seems like the chunk option "eval: false" is not respected, i.e. the code in the chunk in question is nevertheless evaluated.

````markdown
---
title: Test
format: html
---


```{r}
#| eval: false

1 + 1
```

```{r}
2 + 2
```

```{r}
3 + 3
```

````

This works for me - Clicking Run all the above from third chunk will not run the first. 

You may need to update RStudio IDE - I am using last daily version and this may have been fixed recently. Which version are you using ?

**Reply by werkstattcodes** [reply]
Many thanks for getting back to me!

No, I am using VS Code version 1.71,  the VS Code Quarto Extension Quarto  v1.43.0; but in the bottom left corner, it says Quarto 1.1.251 (.....\Programs\Quarto\bin)

![quarto-chunk-eval-falseqmd-Visua](https://user-images.githubusercontent.com/9595585/193402683-12248eaa-53c7-47ed-b570-868174eb1b5a.gif)


However, when I try the code in RStudio IDE, the problem - as you noted - does not occur!

**Reply by eitsupi** [reply]
I think you could try the following methods.

````qmd
```r
1 + 1
```

```{.r}
1 + 1
```

```{{r}}
1 + 1
```
````

**Reply by werkstattcodes** [reply]
Indeed, changing the chunk to any of the three variations mentioned in your reply leads to the chunk not being evaluated when using "run above".  Many thanks. 

Would be nice, though, if eval: false would work like all other chunk options.

**Reply by eitsupi** [reply]
I think there is a tradeoff to be made.
That is, if we want to write a code block that "we don't want executed at render time but will always work when executed", the current behavior of vscode makes sense.

I just don't surround it with `{}` if I want to use just to show code blocks.

**Reply by eitsupi** [reply]
Given the consistency with editing ipynb in Jupyter, I think it makes more sense for the eval option to be ignored by the editor.
(RStudio IDE's behavior maybe need to be changed. Or, have an option to change the behavior at least)

**Comment by cderv** [comment]
FWIW it has been reported in https://github.com/quarto-dev/quarto-vscode/issues/89 and now `eval: false` is respected in VSCODE like in RStudio IDE. (https://github.com/quarto-dev/quarto-vscode/commit/50d648a8a70ea3fb2f79f3357839bbad00d01dd4)

**Reply by werkstattcodes** [reply]
Many thanks. Indeed it works now. There's one thing which I noted now, but I am not sure whether its really an "issue".

If you render the script below, the first chunk will be evaluated, despite being eval:false. The reason is the missing space between | and eval.
Interestingly, though, if i simply trigger the chunk by the subsequent "Run above", it is not triggered (as intended and corrected by the recent change). So, in one instance the space makes a difference, in the other not. Not a big deal if you know it, but it can't be quite annoying if you don't know it...

---
title: Test
format: html
---

```{r}
#| include: false
library(tidyverse)
```

```{r}
#|eval: false
1 + 1
```

```{r}
#| eval: false
2 + 2
```


```{r}
3 + 3
```

**Reply by cderv** [reply]
>  So, in one instance the space makes a difference, in the other not. 

`#|eval: false` should not be valid option comment. It should not be parsed correctly. Space is needed. 
Can you open an issue in https://github.com/quarto-dev/quarto-vscode/ about the Run above different behavior ? 

Thank you

**Comment by jarbet** [comment]
In VSCode, if I click "Run Above" and some of the above cells have `{r, eval = FALSE}` in the header, they will still be run.

However, if I instead put this inside the cell `#| eval: false` then the cell will no longer be run.  So I will use this syntax from now on.

**Reply by mcanouil** [reply]
Quarto uses the YAML syntax.
Knitr supports: R inline, R multiline, and YAML.

A simple rule to remember in Quarto is: in executable code cells, only the language should be in the brackets, nothing else.

**Reply by jarbet** [reply]
> A simple rule to remember in Quarto is: in executable code cells, only the language should be in the brackets, nothing else.

Interesting, I did not know that.  Thanks!
